File: convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
1 2 3 4 5 6
call recursively passing the range of elements 
if range is defined, can create a node. otherwise return false
'''
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def satb(left: int, right: int) -> Optional[TreeNode]:
            if left>right:
                return None
            mid=left+(right-left)//2
            root=TreeNode(nums[mid])
            root.left=satb(left,mid-1)
            root.right=satb(mid+1,right)
            return root
            
        n=len(nums)
        return satb(0,n-1)

File: lc/convert-sorted-array-to-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// recursion: take the middle, call conv on both sides
// if no interval, return nil
// 1 3
// 0 1

// lo hi mi  1 1 1 
//       1
//     /    \
//   nil.    3
func conv(nums []int, lo, hi int) *TreeNode {
    if lo > hi { return nil }
    mi := lo + (hi - lo)/2
    root := &TreeNode{Val: nums[mi]}
    root.Left = conv(nums, lo, mi - 1)
    root.Right = conv(nums, mi + 1, hi)
    return root
}
func sortedArrayToBST(nums []int) *TreeNode {
    return conv(nums, 0, len(nums) - 1)
}